core/geometry/src/curve/internalContexts/geometry2d/ConstraintSet.ts

```diff
@@ -0,0 +1,298 @@
+      c.radius = this.radius;
+    return c;
+  }
+}
+
+export class ConstraintSet {
+  private _constraints: ConstructionConstraint[];
+  private constructor() {
+    this._constraints = [];
+  }
+  /** Create an empty constraint set */
+  public static create(): ConstraintSet {
+    return new ConstraintSet();
+  }
+  /** Clone the array */
+  public clone(): ConstraintSet {
+    const theClone = ConstraintSet.create();
+    for (const c of this._constraints) {
+      theClone._constraints.push(c.clone());
+    }
+    return theClone;
+  }
+
+  /**
+   * Add a new constraint to the array.
+   * @param constraint 
+   * @returns true if the constratnt was accepted.
+   */
+  public addConstraint(constraint: ConstructionConstraint): boolean {
+    this._constraints.push(constraint);
+    return true;
+  }
+  /** Count the constraints of specific type. */
+  public countConstraintType(t: ConstriaintType) {
+    let n = 0;
+    for (const c of this._constraints) {
+      if (c.constraintType === t) n++;
+    }
+    return n;
+  }
+  /** Replace zero radius circle constraints to point constraints. */
+  public convertZeroRadiusCirclesToThroughPoint() {
+    for (let i = 0; i < this._constraints.length; i++) {
+      const c = this._constraints[i];
+      if (c.constraintType === ConstriaintType.tangentTo
+        && c.curve !== undefined
+        && c.curve instanceof UnboundedCircle2dByCenterAndRadius
+        && Geometry.isSmallMetricDistance(c.curve.radius)) {
+        const c1 = ConstructionConstraint.createThroughPoint(c.curve.center);
+        this._constraints[i] = c1;
+      }
+    }
+  }
+  /** Replace point passthrough to zero radius circle tangency */
+  public convertThroughPointToZeroRadiusCircles() {
+    for (let i = 0; i < this._constraints.length; i++) {
+      const c = this._constraints[i];
+      if (c.constraintType === ConstriaintType.throughPoint
+        && c.point !== undefined) {
+        const c1 = ConstructionConstraint.createTangentTo(
+          UnboundedCircle2dByCenterAndRadius.createPointRadius(c.point, 0.0))
+        this._constraints[i] = c1;
+      }
+    }
+  }
+  /** normalize all line perendiculars */
+  public normalizeLines() {
+    for (let i = 0; i < this._constraints.length; i++) {
+      const c = this._constraints[i];
+      if (c.constraintType === ConstriaintType.tangentTo
+        && c.curve instanceof UnboundedLine2dByPointAndNormal) {
+        const newLine = c.curve.cloneNormalizedFromOrigin();
+        if (newLine !== undefined) {
+          const c1 = ConstructionConstraint.createTangentTo(newLine);
+          this._constraints[i] = c1;
+        }
+      }
+    }
+  }
+
+
+  /**
+   * Return
+   *   * 1 if the constraint curve is UnboundedLine2dByPointAndNormal
+   *   * 2 if the constraint curve is UnboundedCircle2dByCenterAndRadius
+   *   * 0 otherwise
+   * @param c constraint
+   */
+  public static tangentGeometryLineCircleType(c: ConstructionConstraint): number {
```

Comment:
what is this function doing? is it just returning a number for each type (in order to sort later) or does the number have a meaning like number of possible tangents to a circle or line? Please clarify the doc.

Reply:
It's a number for sorting.